client/fingerprint_manager.go

```diff
@@ -0,0 +1,122 @@
+	node      *structs.Node
+	client    *Client
+	logger    *log.Logger
+}
+
+// run runs each fingerprinter individually on an ongoing basis
+func (fm *FingerprintManager) run(f fingerprint.Fingerprint, period time.Duration, name string) {
+	fm.logger.Printf("[DEBUG] fingerprint_manager: fingerprinting %s every %v", name, period)
+
+	for {
+		select {
+		case <-time.After(period):
+			_, err := fm.fingerprint(name, f)
+			if err != nil {
+				fm.logger.Printf("[DEBUG] fingerprint_manager: periodic fingerprinting for %v failed: %+v", name, err)
+				continue
+			}
+
+		case <-fm.client.shutdownCh:
+			return
+		}
+	}
+}
+
+// setupDrivers is used to fingerprint the node to see if these drivers are
+// supported
+func (fm *FingerprintManager) SetupDrivers(drivers []string) error {
+	var availDrivers []string
+	driverCtx := driver.NewDriverContext("", "", fm.getConfig(), fm.node, fm.logger, nil)
+	for _, name := range drivers {
+
+		d, err := driver.NewDriver(name, driverCtx)
+		if err != nil {
+			return err
+		}
+
+		detected, err := fm.fingerprint(name, d)
+		if err != nil {
+			fm.logger.Printf("[DEBUG] fingerprint_manager: fingerprinting for %v failed: %+v", name, err)
+			return err
+		}
+
+		// log the fingerprinters which have been applied
+		if detected {
+			availDrivers = append(availDrivers, name)
+		}
+
+		p, period := d.Periodic()
+		if p {
+			go fm.run(d, period, name)
+		}
+	}
+
+	fm.logger.Printf("[DEBUG] fingerprint_manager: available drivers %v", availDrivers)
+
+	return nil
+}
+
+// fingerprint does an initial fingerprint of the client. If the fingerprinter
+// is meant to be run continuously, a process is launched ro perform this
+// fingerprint on an ongoing basis in the background.
+func (fm *FingerprintManager) fingerprint(name string, f fingerprint.Fingerprint) (bool, error) {
+	request := &cstructs.FingerprintRequest{Config: fm.getConfig(), Node: fm.node}
+	var response cstructs.FingerprintResponse
+	err := f.Fingerprint(request, &response)
+	if err != nil {
+		return false, err
+	}
+
+	fm.client.updateNodeFromFingerprint(&response)
+	return response.Detected, nil
+}
+
+// setupDrivers is used to fingerprint the node to see if these attributes are
+// supported
+func (fm *FingerprintManager) SetupFingerprints(fingerprints []string) error {
+	var appliedFingerprints []string
+
+	for _, name := range fingerprints {
+		f, err := fingerprint.NewFingerprint(name, fm.logger)
+
+		if err != nil {
+			fm.logger.Printf("[DEBUG] fingerprint_manager: fingerprinting for %v failed: %+v", name, err)
+			return err
+		}
+
+		detected, err := fm.fingerprint(name, f)
+		if err != nil {
+			return err
+		}
+
+		// log the fingerprinters which have been applied
+		if detected {
+			appliedFingerprints = append(appliedFingerprints, name)
+		}
+
+		p, period := f.Periodic()
+		if p {
+			go fm.run(f, period, name)
+		}
+	}
+
+	fm.logger.Printf("[DEBUG] fingerprint_manager: applied fingerprints %v", appliedFingerprints)
```

Comment:
shouldn;t this be applied fingerprinters in the log message?

Reply:
The type is called "fingerprints" - I'm ok with either in the log message, but I don't think we should do a full rename of the type.

Reply:
I think the log message should be accurate - it was a source of confusion in a previous review too. Would be fine not renaming the type since that's not visible to users.